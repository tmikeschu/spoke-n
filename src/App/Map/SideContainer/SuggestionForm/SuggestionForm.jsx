import React, { Component } from "react"
import PropTypes from 'prop-types'
import APIService from "../../../APIService/APIService"
import { Notification } from "react-notification"

const service = new APIService("https://spoken-api.herokuapp.com")

export default class SuggestionForm extends Component {
  constructor(props) {
    super(props)
    this.state = {
      pin: {
        name: "",
        email: "",
        label: "",
        description: "",
        category: "",
        message: "",
      },
      suggestionSent: false,
      suggestionFailed: false,
      formNotification: "",
    }
    this.clearedForm = this.state.pin
  }

  handleChange = event => {
    this.setState({
      pin: {
        ...this.state.pin,
        [event.target.name]: event.target.value
      }
    })
  }

  handleSubmit = async event => {
    event.preventDefault();
    const pin = {
      pin: {
        ...this.state.pin,
        lat: this.props.suggestionPin.lat,
        lng: this.props.suggestionPin.lng
      }
    }
    try {
      await service.post("/api/v1/suggestion_pins", pin)
      this.submitSuccess()
    } catch(error) {
      this.submitFail(error)
    }
  }

  submitSuccess() {
    this.props.getSuggestions()
    this.props.setSuggestion({})
    this.setState({
      formNotification: "Suggestion sent!",
      suggestionSent: true,
      pin: this.clearedForm
    })
  }

  submitFail(error) {
    this.setState({
      formNotification: error.responseJSON["error"],
      suggestionFailed: true
    })
  }

  deactivate() {
    this.setState({
      suggestionSent: false,
      suggestionFailed: false,
    })
  }

  render() {
    const { suggestionPin } = this.props

    const form = (
      <form onSubmit={this.handleSubmit}>
        <input required type="text" name="name" value={this.state.pin.name}
          onChange={this.handleChange} placeholder="Name" />

        <input required type="email" name="email" value={this.state.pin.email}
          onChange={this.handleChange} placeholder="Email (our eyes only!)" />

        <input required type="text" name="label" value={this.state.pin.label}
          onChange={this.handleChange} placeholder="Label" />

        <textarea required name="description" value={this.state.pin.description}
          onChange={this.handleChange} placeholder="Description" />

        <select required name="category" value={this.state.pin.category}
          onChange={this.handleChange} >
          <option value="">--Category--</option>
          <option value="0">Place to stay</option>
          <option value="1">Cool spot</option>
          <option value="2">Avoid this place</option>
          <option value="3">Bike shop</option>
          <option value="4">Other</option>
        </select>

        <textarea name="message" value={this.state.pin.message}
          onChange={this.handleChange} placeholder="Message (optional)" />

        <p style={{
            display: suggestionPin.lat === undefined ? "block" : "none"
          }}
          value="Drop It!"
        >
          Click the map to drop a pin!
        </p>

        <input type={
            suggestionPin.lat === undefined ? "hidden" : "submit"
          }
          value="Drop It!" />

      </form>
    )
    return (
      <article className="suggestion-form" >
        <h4>Drop a Pin</h4>
        { form }
        <Notification
          isActive={this.state.suggestionSent}
          message={this.state.formNotification}
          onClick={() => this.deactivate() }
          action="âœ–"
        />
      </article>
    )
  }
}

SuggestionForm.propTypes = {
  suggestionPin: PropTypes.object.isRequired,
  getSuggestions: PropTypes.func.isRequired,
  setSuggestion: PropTypes.func.isRequired
}

