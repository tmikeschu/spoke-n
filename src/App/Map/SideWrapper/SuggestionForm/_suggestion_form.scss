.SuggestionForm {
  background: rgba(0, 0, 0, 0.1);
  margin: 0 -8px -8px;
  padding: 8px;
  h4 {
    margin-bottom: 8px;
  }
  p {
    font-style: italic;
  }
  form {
    display: flex;
    flex-flow: column nowrap;
    textarea,
    input,
    select {
      margin: 8px 0;
      border: none;
      font-family: $body-font;
    }
    textarea {
      resize: none;
      height: 64px;
    }
  }
  p {
    text-align: center;
    width: 100%;
  }
}
