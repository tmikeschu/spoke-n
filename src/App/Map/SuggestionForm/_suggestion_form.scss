.suggestion-form {
  width: 99%;
  background: rgba(0,0,0,.1);
  border: 2px dotted $yellow;
  color: $green;
  padding: 5px 0;
  h4 {
    text-align: center;
  }
  p {
    font-style: italic;
  }
  form {
    display: flex;
    flex-flow: column nowrap;
    width: 95%;
    margin: 0 auto;
    textarea, input, select {
      margin: 3% 0;
      width: 95;
      font-size: 1.2em;
      border: none;
      font-family: $body-font;
    }
    textarea {
      resize: none;
      height: 4em;
    }
    input {
      &[type="submit"] {
        width: 25%;
        align-self: center;
        background-color: $lilac;
        color: white;
        &:hover, &:focus {
          background-color: $blue;
        }
      }
    }
  }
  p {
    text-align: center;
    width: 100%;
  }
  @media (max-width: 640px) {
    width: 100%;
    font-size: 16px;
    h4 {
      font-size: 1em;
    }
    input, textarea, select {
      font-size: .5em;
    }
    p {
      font-size: .9em;
    }
    form {
      font-size: .5em;
      width: 90%;
      input, textarea {
        border: 1px solid #aaa;
      }
      input {
        &[type="submit"] {
          width: 100%;
        }
      }
    }
  }
}

