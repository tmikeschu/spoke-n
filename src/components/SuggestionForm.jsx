import React, { Component } from 'react';
import $ from 'jquery';
import { Notification } from 'react-notification';
import '../stylesheets/App.css';

export default class SuggestionForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      label: '',
      description: '',
      category: '',
      message: '',
      suggestionSent: false,
      suggestionFailed: false,
      formNotification: '',
    };

    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleChange(event) {
    this.setState({
      [event.target.name]: event.target.value
    });
  }

  handleSubmit(event) {
    event.preventDefault();
    let self = this;
    $.ajax({
      url: "http://spoken-api.herokuapp.com/api/v1/suggestion_pins?api_key="+process.env.REACT_APP_RAILS_KEY,
      method: "POST",
      data: {
        pin: {
          label: this.state.label,
          description: this.state.description,
          category: this.state.category,
          lat: this.props.suggestionPin.lat,
          lng: this.props.suggestionPin.lng,
          message: this.state.message,
        }
      },
      dataType: "json"
    }).done(function(data) {
      self.props.getSuggestions();
      self.props.setSuggestion({});
      self.setState({
        formNotification: "Suggestion sent!",
        suggestionSent: true,
      });
    }).fail(function(error) {
      self.setState({
        formNotification: error.responseJSON["error"],
        suggestionFailed: true,
      });
    });
  }

  deactivate() {
    this.setState({
      suggestionSent: false,
      suggestionFailed: false,
    })
  }

  render() {
    return (
      <article className="suggestion-form">
        <form onSubmit={this.handleSubmit}>
          <input required type="text" name="label" value={this.state.label} onChange={this.handleChange} placeholder="Label" />
          <textarea required name="description" value={this.state.description} onChange={this.handleChange} placeholder="Description"/>
          <select required name="category" value={this.state.category} onChange={this.handleChange} >
            <option value="">--Category--</option>
            <option value="0">Place to stay</option>
            <option value="1">Cool spot</option>
            <option value="2">Avoid this place</option>
            <option value="3">Bike shop</option>
            <option value="4">Other</option>
          </select>
          <textarea name="message" value={this.state.message} onChange={this.handleChange} placeholder="What's up?"/>
          <input type={this.props.suggestionPin.lat === undefined ? "hidden" : "submit"} value="Drop It!" />
        </form>
        <p style={{display: this.props.suggestionPin.lat === undefined ? "block" : "none"}}>Click the suggestion map to drop a pin!</p>
        <Notification
          isActive={this.state.suggestionSent}
          message={this.state.formNotification}
          onClick={() => this.deactivate() }
          action="âœ–"
        />
      </article>
    );
  }
}

