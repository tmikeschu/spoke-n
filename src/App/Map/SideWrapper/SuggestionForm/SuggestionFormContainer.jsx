import { connect } from "react-redux"
import { bindActionCreators } from "redux"
import * as actions from "../../../../Redux/actions"
import SuggestionForm from "./SuggestionForm"

const mapStateToProps = ({ map }) => map

const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators(actions, dispatch)
})

export default connect(mapStateToProps, mapDispatchToProps)(SuggestionForm)

