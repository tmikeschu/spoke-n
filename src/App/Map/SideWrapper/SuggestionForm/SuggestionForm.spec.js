import React from "react"
import { shallow } from "enzyme"
import SuggestionForm from "./SuggestionForm"

const props = {
  suggestionPin: {},
  actions: {
    addSuggestionPin: jest.fn(),
    addSuggestions: jest.fn(),
  }
}

describe("<SuggestionForm />", () => {
  const wrapper = shallow(<SuggestionForm {...props} />)
  const suggestionForm = wrapper.instance()

  it("renders without crashing", () => {
    expect(wrapper).toBeTruthy()
  })

  describe("snapshot", () => {
    it("is valid", () => {
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe("initial state", () => {
    it("is valid", () => {
      const expectedState = { 
        pin: { 
          name: '',
          email: '',
          label: '',
          description: '',
          category: '',
          message: '' 
        },
        suggestionSent: false,
        suggestionFailed: false,
        formNotification: '' 
      }
      expect(suggestionForm.state).toMatchObject(expectedState)
    })
    it("has an empty form for resets", () => {
      const emptyForm = { 
        name: '',
        email: '',
        label: '',
        description: '',
        category: '',
        message: '' 
      }
      expect(suggestionForm.clearedForm).toMatchObject(emptyForm)
    })
  })

  describe("handleChange", () => {
    it("sets state for pin information", () => {
      expect(suggestionForm.state.pin.email).toEqual("")
      suggestionForm.handleChange(fakeEvent)
      expect(suggestionForm.state.pin.email).toEqual("tommy@crosby.com")
    })
  })

  describe("handleSubmit", () => {
    it("calls postSuggestion", () => {
      const restore = suggestionForm.postSuggestion
      const mock = suggestionForm.postSuggestion = jest.fn()

      suggestionForm.handleSubmit(fakeEvent)
      expect(mock).toHaveBeenCalled()
      suggestionForm.postSuggestion = restore
    })
  })

  describe("postSuggestion", () => {
    it("calls service.post", async () => {
      const restore = fakeService.post
      const mock = fakeService.post = jest.fn()

      await suggestionForm.postSuggestion(fakeService, {})
      expect(mock).toHaveBeenCalled()
      fakeService.post = restore
    })

    it("calls submitSuccess", async () => {
      const restore = suggestionForm.submitSuccess
      const mock = suggestionForm.submitSuccess = jest.fn()
      await suggestionForm.postSuggestion(fakeService, {})

      expect(mock).toHaveBeenCalled()
      suggestionForm.submitSuccess = restore
    })
  })

  describe("submitSuccess", () => {
    const actions = ['addSuggestions', 'addSuggestionPin']
    actions.forEach(action => {
      it(`calls the ${action} action`, async () => {
        const mock = suggestionForm.props.actions[action]
        await suggestionForm.submitSuccess()
        expect(mock).toHaveBeenCalled()
        mock.mockReset()
      })
    })

    it("updates form state", async () => {
      const suggestionForm = shallow(<SuggestionForm {...props} />).instance()
      const newState = {
        pin: {
          name: "Name",
          email: "YES",
          label: 'Something',
          description: 'Desc',
          category: 'Stay',
          message: 'hello'
        },
        suggestionSent: false,
        suggestionFailed: false,
        formNotification: ''
      }
      suggestionForm.setState(newState)

      expect(suggestionForm.state).toMatchObject(newState)

      await suggestionForm.submitSuccess()

      expect(suggestionForm.state).not.toMatchObject(newState)
      expect(suggestionForm.state).toMatchObject({
        formNotification: "Suggestion sent!",
        suggestionSent: true,
        pin: { ...suggestionForm.clearedForm }
      })
    })
  })

  describe("submitFail", () => {
    it("updates state correctly", () => {
      const suggestionForm = shallow(<SuggestionForm {...props} />).instance()
      const error = {
        responseJSON: { error: "Failed!" }
      }
      const newState = {
        formNotification: "Failed!",
        suggestionFailed: true
      }

      expect(suggestionForm.state).not.toMatchObject(newState)
      suggestionForm.submitFail(error)
      expect(suggestionForm.state).toMatchObject(newState)
    })
  })

  describe("deactivate", () => {
    it("resets state for sent/failed", () => {
      suggestionForm.setState({ suggestionSent: true, suggestionFailed: true, })
      const deactivated = { suggestionSent: false, suggestionFailed: false, }

      expect(suggestionForm.state).not.toMatchObject(deactivated)
      suggestionForm.deactivate()
      expect(suggestionForm.state).toMatchObject(deactivated)
    })
  })
})

const fakeEvent = {
  preventDefault () {},
  target: {
    name: "email",
    value: "tommy@crosby.com"
  }
}

const fakeService = {
  post (url, data) {}
}
