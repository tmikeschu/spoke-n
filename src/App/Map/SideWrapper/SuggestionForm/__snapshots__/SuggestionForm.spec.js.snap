// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SuggestionForm /> snapshot is valid 1`] = `
<article
  className="SuggestionForm"
>
  <h4>
    Drop a Pin 
    <span
      aria-label="pin"
      role="img"
    >
      ðŸ“Œ
    </span>
  </h4>
  <form
    onSubmit={[Function]}
  >
    <input
      name="name"
      onChange={[Function]}
      placeholder="Name"
      required={true}
      type="text"
      value=""
    />
    <input
      name="email"
      onChange={[Function]}
      placeholder="Email (our eyes only!)"
      required={true}
      type="email"
      value=""
    />
    <input
      name="label"
      onChange={[Function]}
      placeholder="Label"
      required={true}
      type="text"
      value=""
    />
    <textarea
      name="description"
      onChange={[Function]}
      placeholder="Description"
      required={true}
      value=""
    />
    <select
      name="category"
      onChange={[Function]}
      required={true}
      value=""
    >
      <option
        value=""
      >
        --Category--
      </option>
      <option
        value="0"
      >
        Place to stay
      </option>
      <option
        value="1"
      >
        Cool spot
      </option>
      <option
        value="2"
      >
        Avoid this place
      </option>
      <option
        value="3"
      >
        Bike shop
      </option>
      <option
        value="4"
      >
        Other
      </option>
    </select>
    <textarea
      name="message"
      onChange={[Function]}
      placeholder="Message (optional)"
      value=""
    />
    <p
      style={
        Object {
          "display": "block",
        }
      }
      value="Drop It!"
    >
      Click the map to drop a pin!
    </p>
    <input
      type="hidden"
      value="Drop It!"
    />
  </form>
  <Notification
    action="âœ–"
    activeClassName="notification-bar-active"
    dismissAfter={2000}
    isActive={false}
    message=""
    onClick={[Function]}
  />
</article>
`;
